lib/command.js

```diff
@@ -1249,6 +1233,83 @@ Expecting one of '${allowedValues.join("', '")}'`);
     return result;
   }
 
+  /**
+   * Call parseArgs with extra handling:
+   * - custom error message if parseArgs throws 'commander.invalidArgument'
+   * - if the previous value is a promise, chain the call to resolve the promise before parsing
+   *
+   * @param {Option | Argument} target
+   * @param {string} value
+   * @param {Promise<any> | any} previous
+   * @param {string} invalidArgumentMessage
+   * @api private
+   */
+
+  _parseArg(target, value, previous, invalidArgumentMessage) {
+    const refineError = (err) => {
+      if (err.code === 'commander.invalidArgument') {
+        const message = `${invalidArgumentMessage} ${err.message}`;
+        this.error(message, { exitCode: err.exitCode, code: err.code });
+      }
+      throw err;
+    };
+
+    let result;
+    if (thenable(previous)) {
+      result = previous.then(resolvedPrevious => {
+        let innerResult = target.parseArg(value, resolvedPrevious);
+        if (thenable(innerResult)) innerResult = innerResult.then(null, refineError); // .catch
+        return innerResult;
+      });
+    } else {
+      try {
+        result = target.parseArg(value, previous);
+        if (thenable(result)) result = result.then(null, refineError); // .catch
```

Comment:
`result` is going to be the next `previous` value, so for `resolvedPrevious` to be correct, `x => x` has to be used instead of `null`.

In #19115, I also went on to use `_chainOrCall()` to follow the DRY principle.